### PR TITLE
[BUG] multi-thread flusher causes data inconsistent with chunk offsets

### LINKED ISSUE #274: [BUG] Multi-thread flusher causes data inconsistent with chunk offsets.
### How to reproduce the bug?
This problem happens randomly. There are no clear procedures to reproduce.

### Could you share logs or screenshots?

> Caused by: java.lang.IndexOutOfBoundsException: readerIndex(25350231) + length(30480) exceeds writerIndex(25358446): UnpooledDuplicatedByteBuf(ridx: 25350231, widx: 25358446, cap: 25358446, unwrapped: CompositeByteBuf(ridx: 13, widx: 25358446, cap: 25358446, components=454))
>   at com.aliyun.emr.io.netty.buffer.AbstractByteBuf.checkReadableBytes0(AbstractByteBuf.java:1442)

### PR DESCRIPTION
**Pull request by FMX**
# [BUG] multi-thread flusher causes data inconsistent with chunk offsets

### Related issues.
closes #274 

### How was this patch tested?

/cc @related-reviewer

/assign @main-reviewer

### CODE REVIEW

File: server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/FileWriter.java [hunk 1]
```diff
@@ -61,6 +61,7 @@ public final class FileWriter extends DeviceObserver {
   private long bytesFlushed;
 
   private final DiskFlusher flusher;
+  private final int flusherReplicationIndex;
```
Comment (waitinfuture) [inline server-worker/src/main/java/com/aliyun/emr/rss/service/deploy/worker/FileWriter.java:64]: flushWorkerIndex

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala [hunk 2]
```diff
@@ -53,13 +53,22 @@ private[worker] final class DiskFlusher(
...
-  private val workingQueue = new LinkedBlockingQueue[FlushTask](queueCapacity)
-  private val bufferQueue = new LinkedBlockingQueue[CompositeByteBuf](queueCapacity)
-  private val writeActionPool = ThreadUtils.newDaemonFixedThreadPool(threadCount,
-    workingDir.getName + "-flusher")
-  for (_ <- 0 until queueCapacity) {
-    bufferQueue.put(Unpooled.compositeBuffer(256))
+  private val workingQueues = new Array[LinkedBlockingQueue[FlushTask]](threadCount)
+  private val bufferQueues = new Array[LinkedBlockingQueue[CompositeByteBuf]](threadCount)
+  for (index <- 0 until threadCount) {
+    val actualQueueSize = if (queueCapacity / threadCount > 0) {
```
Comment (AngersZhuuuu) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala:59]: Maybe it's better to add some assert conditions here? or  add this part to RssConf?

Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala:59]: Simplify to queueCapacity / threadCount + 1

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala [hunk 3]
```diff
@@ -90,55 +101,60 @@ private[worker] final class DiskFlusher(
...
       case e: Exception =>
-        logError(s"Exception when interrupt worker: $worker, $e")
+        logError(s"Exception when interrupt worker: $workers, $e")
+    }
+    workingQueues.foreach { queue =>
+      queue.asScala.foreach(task => {
+        task.buffer.removeComponents(0, task.buffer.numComponents())
+        task.buffer.clear()
+      })
     }
```
Comment (AngersZhuuuu) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala:157]:
```
 workingQueues.foreach { queue =>
      queue.asScala.foreach { task =>
        task.buffer.removeComponents(0, task.buffer.numComponents())
        task.buffer.clear()
      }
    }
```

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala [hunk 4]
```diff
@@ -53,13 +53,22 @@ private[worker] final class DiskFlusher(
...
+      100
+    }
+    workingQueues(index) = new LinkedBlockingQueue[FlushTask](actualQueueSize)
+    bufferQueues(index) = new LinkedBlockingQueue[CompositeByteBuf](actualQueueSize)
+    for (_ <- 0 until actualQueueSize) {
+      bufferQueues(index).put(Unpooled.compositeBuffer(256))
+    }
   }
+  private val workers = new Array[Thread](threadCount)
+  private val nextReplicationIndex = new AtomicInteger()
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala:71]: We should protect from overflow

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala [hunk 5]
```diff
@@ -90,55 +101,60 @@ private[worker] final class DiskFlusher(
...
-  worker.start()
 
-  deviceMonitor.registerDiskFlusher(this)
+  def getReplicationIndex: Int = {
+    nextReplicationIndex.getAndIncrement() % threadCount
+  }
 
-  def takeBuffer(timeoutMs: Long): CompositeByteBuf = {
-    bufferQueue.poll(timeoutMs, TimeUnit.MILLISECONDS)
+  def takeBuffer(timeoutMs: Long, replicationIndex: Int): CompositeByteBuf = {
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala:126]: workerIndex, and others

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala [hunk 6]
```diff
@@ -147,7 +163,7 @@ private[worker] final class DiskFlusher(
     deviceMonitor.reportDeviceError(workingDir, e, deviceErrorType)
   }
 
-  def bufferQueueInfo(): String = s"$this available buffers: ${bufferQueue.size()}"
+  def bufferQueueInfo(): String = s"$this used buffers: ${bufferQueues.map(_.size()).toList}"
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala:166]: used => available
${bufferQueues.map(_.size()).toList => ${bufferQueues.map(_.size()).sum

**Review by AngersZhuuuu — COMMENTED** [review]
LGTM

**Comment by FMX** [comment]
> LGTM, one thought is if we can combine `workingQueue`, `bufferQueue`, `workers` to one class?

I think it's a good idea.

File: server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala [hunk 7]
```diff
@@ -90,55 +97,64 @@ private[worker] final class DiskFlusher(
...
-    override def uncaughtException(t: Thread, e: Throwable): Unit = {
-      logError(s"$this thread terminated.", e)
-    }
-  })
-  worker.start()
 
-  deviceMonitor.registerDiskFlusher(this)
+  def getWorkerIndex: Int = {
+    val nextIndex = nextWorkerIndex.getAndIncrement()
+    if (nextIndex > 1000000) {
```
Comment (waitinfuture) [inline server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala:120]: 1000000 => threadCount

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

**Pull request merged**